validate-xml/validate-xml.cpp

```diff
@@ -0,0 +1,372 @@
+             << "\": " << StrX(toCatch.getMessage()) << endl;
+    }
+    void fatalError(const SAXParseException& toCatch) {
+        cerr << "Fatal Error at file \"" << StrX(toCatch.getSystemId())
+             << "\", line " << toCatch.getLineNumber()
+             << ", column " << toCatch.getColumnNumber() << endl
+             << "Message: " << StrX(toCatch.getMessage()) << endl;
+    }
+};
+static bool IsFileExists(const char* filePath)
+{
+    struct stat fileStat;
+    if (stat(filePath, &fileStat) == -1 && errno == ENOENT)
+        return false;
+    return true;
+}
+//---------------------------------------------------------------------------
+void printUsage()
+{
+	cout<<"Usage: XMLValidation <XML File Path>"<<endl;
+	cout<<"Exits with code"<<endl;
+	cout<<"0 - Valid XML"<<endl<<"1 - Schema Check Error"<<endl<<"2 - Signature Validation Error"<<endl;
+}
+
+int IsFilePathValid(const char* filename)
+{
+	if(IsFileExists(filename))
+	{
+		string fn = filename;
+		if(fn.substr(fn.find_last_of(".") + 1) == "xml")
+			return 1;
+		else
+			return 0;
+	}
+	else
+		return 0;
+}
+
+int IsSigValid(DOMDocument *theDOM, DOMNode *sigNode)
+{
+	XSECProvider prov;
+	XSECKeyInfoResolverDefault theKeyInfoResolver;
+	DSIGSignature * sig = prov.newSignatureFromDOM(theDOM, sigNode);
+
+	sig->setKeyInfoResolver(&theKeyInfoResolver);
+	bool result;
+	try
+	{
+		sig->load();
+		result = sig->verifySignatureOnly();
+	}
+	catch (XSECException &e)
+	{
+		char * msg = XMLString::transcode(e.getMsg());
+		cerr << "An error occured during signature verification\n   Message: "
+		<< msg << endl;
+		XSEC_RELEASE_XMLCH(msg);
+		return 0;
+	}
+	catch (XSECCryptoException &e)
+	{
+		cerr << "An error occured during signature verification\n   Message: "
+		<< e.getMsg() << endl;
+		return 0;
+	}
+	catch (...) 
+	{
+		cerr << "Unknown Exception type occured.  Cleaning up and exiting\n" << endl;
+		return 0;
+	}
+	if (result)
+	{
+		cout << "Valid XML" << endl;
+		return 1;
+	}
+	else
+	{
+		cout << "Invalid XML" << endl;
+		char * e = XMLString::transcode(sig->getErrMsgs());
+		cout << e << endl;
+		XSEC_RELEASE_XMLCH(e);
+		return 0;
+	}
+	
+}
+	
+
+int main(int argc, const char** argv)
+{
+	if (argc < 2)
+	{
+		printUsage();
+		exit (1);
+	}
+	else
+	{
+		if(!IsFilePathValid(argv[1]))
+		{
+		cout<<"Invalid File";
+		exit(1);
+		}
+	}
+    try
+    {
+       XMLPlatformUtils::Initialize();
+	   XSECPlatformUtils::Initialise();
+    }
+    catch(const XMLException& e)
+    {
+        StrX tmp_e(e.getMessage());
+        cerr << "Initialization error: " << tmp_e.localForm() << endl;
+        return schemaError;
+    }
+
+    unique_ptr<XercesDOMParser> parser(new XercesDOMParser);
+    unique_ptr<DOMTreeErrorReporter> errReporter(new DOMTreeErrorReporter());
+    parser->setErrorHandler(errReporter.get());
+    parser->setDoNamespaces(true);
+    parser->useCachedGrammarInParse(true);
+    parser->setValidationScheme(XercesDOMParser::Val_Auto);
```

Comment:
@RajaBoseLeoM In the Xerces docs it said parser->cacheGrammarFromParse(true) is required only if we parse the document more than once. Here we are parsing it only once, so do you still want me to add it?